airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3StorageOperations.java

```diff
@@ -0,0 +1,172 @@
+    if (!s3Client.doesBucketExistV2(bucket)) {
+      LOGGER.info("Bucket {} does not exist; creating...", bucket);
+      s3Client.createBucket(bucket);
+      LOGGER.info("Bucket {} has been created.", bucket);
+    }
+    if (!s3Client.doesObjectExist(bucket, bucketName)) {
+      LOGGER.info("Storage Object {}/{} does not exist in bucket; creating...", bucket, bucketName);
+      s3Client.putObject(bucket, bucketName.endsWith("/") ? bucketName : bucketName + "/", "");
+      LOGGER.info("Storage Object {}/{} has been created in bucket.", bucket, bucketName);
+    }
+  }
+
+  @Override
+  public String uploadRecordsToBucket(final RecordBufferImplementation recordsData, final String namespace, final String bucketPath)
+      throws Exception {
+    final List<Exception> exceptionsThrown = new ArrayList<>();
+    boolean succeeded = false;
+    while (exceptionsThrown.size() < UPLOAD_RETRY_LIMIT && !succeeded) {
+      try {
+        loadDataIntoBucket(bucketPath, recordsData);
+        succeeded = true;
+      } catch (final Exception e) {
+        LOGGER.error("Failed to upload records into storage {}", bucketPath, e);
+        exceptionsThrown.add(e);
+      }
+      if (!succeeded) {
+        LOGGER.info("Retrying to upload records into storage {} ({}/{}})", bucketPath, exceptionsThrown.size(), UPLOAD_RETRY_LIMIT);
+        // Force a reconnection before retrying in case error was due to network issues...
+        s3Client = s3Config.resetS3Client();
+      }
+    }
+    if (!succeeded) {
+      throw new RuntimeException(String.format("Exceptions thrown while uploading records into storage: %s", Strings.join(exceptionsThrown, "\n")));
+    }
+    return recordsData.getFilename();
+  }
+
+  private void loadDataIntoBucket(final String bucketPath, final RecordBufferImplementation recordsData) throws IOException {
+    final long partSize = s3Config.getFormatConfig() != null ? s3Config.getFormatConfig().getPartSize() : DEFAULT_PART_SIZE;
+    final String bucket = s3Config.getBucketName();
+    final String objectKey = String.format("%s%s", bucketPath, recordsData.getFilename());
+    final StreamTransferManager uploadManager = S3StreamTransferManagerHelper
+        .getDefault(bucket, objectKey, s3Client, partSize)
+        .checkIntegrity(true)
+        .numUploadThreads(DEFAULT_UPLOAD_THREADS)
+        .queueCapacity(DEFAULT_QUEUE_CAPACITY);
+    boolean hasFailed = false;
+    try {
+      final List<MultiPartOutputStream> outputStreams = uploadManager.getMultiPartOutputStreams();
+      try (final MultiPartOutputStream outputStream = outputStreams.get(0)) {
+        try (final InputStream dataStream = recordsData.getInputStream()) {
+          dataStream.transferTo(outputStream);
+        }
+      }
+    } catch (final Exception e) {
+      LOGGER.error("Failed to load data into storage {}", bucketPath, e);
+      hasFailed = true;
+      throw new RuntimeException(e);
+    } finally {
+      if (hasFailed) {
```

Comment:
is there a benefit to having a finally block instead of 
```
try {
        <...>
        uploadManager.complete();
} catch (final Exception e) {
      LOGGER.error("Failed to load data into storage {}", bucketPath, e);
        uploadManager.abort();
       throw new RuntimeException(e);
}

Reply:
We've had issues where the `uploadManager` is never called with neither `.abort()` nor `.complete()` in the past. (due to some Exceptions (or Throwable?) not being caught?)

This resulted in threads being left open and idle and holding the JVM from exiting and thus a stuck sync.

So that's why I've been doing finally blocks instead...